test/teststl.cpp

```diff
@@ -4411,6 +4413,79 @@ class TestStl : public TestFixture {
               true);
         ASSERT_EQUALS("", errout.str());
     }
+
+    void checkMutexes() {
+        check("void f() {\n"
+              "    static std::mutex m;\n"
+              "    static std::lock_guard<std::mutex> g(m);\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:3]: (warning) The lock guard won't unlock until the end of the program which could lead to a deadlock.\n", errout.str());
+
+        check("void f() {\n"
+              "    static std::mutex m;\n"
+              "    std::lock_guard<std::mutex> g(m);\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("", errout.str());
+
+        check("void f() {\n"
+              "    std::mutex m;\n"
+              "    std::lock_guard<std::mutex> g(m);\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:3]: (warning) The lock is ineffective because the mutex is locked at the same scope as the mutex itself.\n", errout.str());
+
+        check("void g();\n"
+              "void f() {\n"
+              "    static std::mutex m;\n"
+              "    m.lock();\n"
+              "    g();\n"
+              "    m.unlock();\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("", errout.str());
+
+        check("void g();\n"
+              "void f() {\n"
+              "    std::mutex m;\n"
+              "    m.lock();\n"
+              "    g();\n"
+              "    m.unlock();\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:4]: (warning) The lock is ineffective because the mutex is locked at the same scope as the mutex itself.\n", errout.str());
```

Comment:
Can't there be some valid cases.. maybe something like this:

    std::mutex m;
    std::thread t([&m](){
        m.lock();
        foo();
        m.unlock();
    });
    m.lock();
    bar();
    m.unlock();

Reply:
That code is prone to crashing since a reference to a local variable is passed into a thread. You need to at least wait for the thread to finish or should use a `std::shared_ptr` for it.

I think having a `std::mutex` object locally never makes any sense since it needs to be available across multiple threads.

Reply:
Well there could be a .join() called before the function finishes. Although the code does seem odd.

Reply:
I can only guess. Imagine a function has two loops. and the developer wants to run those in parallell.

I don't have a strong opinion. I do not use threads much at all. We could merge this and then look at what warnings this generate.